### ISSUE TITLE
upstream `apex.normalization.FusedRMSNorm`

### ISSUE THREAD

**Issue by stas00** [pytorch/pytorch#72643]
### 🚀 The feature, motivation and pitch

All T5 models and their derivatives (t5, mt5, t0, etc.) use `RMSNorm`, instead of `LayerNorm`. The former is like the latter, but it doesn't shift - only scales.

The original need was a discovery that all HF Transformers t5-based models were somewhat slow under mixed precision, because of "manual" implementation of `T5LayerNorm` where manual up/down- casting was causing a significant bottleneck.

While researching this I have run into other users who wanted to use a fast `RMSNorm` (but didn't save the references)

NVIDIA/apex [recently implemented `apex.normalization.FusedRMSNorm`](https://github.com/NVIDIA/apex/pulls?q=is%3Apr+FusedRMSNorm+is%3Aclosed) but building apex is far from easy for a lay person. 

So if possible it'd be great to have this sub-set functionality of `LayerNorm` available in pytorch.

Thank you!

**Comment by vadimkantorov** [comment]
(if it's a variant of LayerNorm, should it be supported by LayerNorm natively, e.g. if weight is not None and bias is None, it should call this new fused kernel?)

**Comment by stas00** [comment]
From https://arxiv.org/abs/1910.07467

> When the mean of summed inputs is zero, RMSNorm is exactly equal to LayerNorm. 

![snapshot_10](https://user-images.githubusercontent.com/10676103/153499706-23b7daa2-161e-4039-a919-b804340bacd8.png)

So it's the same as LayerNorm, but:

1. mean is always 0.
2. no bias - only weight

**Comment by albanD** [comment]
> if it's a variant of LayerNorm, should it be supported by LayerNorm natively, e.g. if weight is not None and bias is None, it should call this new fused kernel?

@vadimkantorov the problem is that you also want to ignore the average term there?
So that would need to be a new boolean flag to enable that. Not sure if that is better than adding a new `RMSNorm` function.

**Comment by PetrochukM** [comment]
It'd be awesome to have this available in PyTorch!!

**Comment by thomasw21** [comment]
Another big model using it: https://arxiv.org/abs/2302.13971 
@albanD I can help add it if necessary, but in terms of design choices which is better:
 - adding boolean flag to enable them in LayerNorm?
 - Creating a new `RMSNorm` class?

**Comment by albanD** [comment]
Hey!
Sorry for the delay in answering, it is a tricky one.
Given the similarity with LayerNorm I think we want to move forward as follows:
- Add a `no_bias` kwarg to LayerNorm which can be used to enable this behavior.
- Make sure the doc for this arg mentions RMSNorm explicitly so that it is discoverable by that name.
- Using torch.compile should gives you the same performance as the fused implementation from apex. So we don't want to add the fused implementation in PyTorch.

**Comment by stas00** [comment]
`torch.compile` is very very very very far from being ready for general use. Most of my attempts at using it failed. You can see multiple bug reports I have filed about it.

But if this request is put on hold for another year it might come through.

**Comment by PetrochukM** [comment]
`torch.compile` along with `JIT` before it has historically not worked well for NLP models which tend to be more dynamic. It'd be nice to prioritize some core development until compilers like that are stabilized.

**Comment by thomasw21** [comment]
Just fyi, RMSNorm isn't just LayerNorm without bias (otherwise we could have use the function method that allows to not pass bias). You need also to remove mean estimation.

**Comment by vadimkantorov** [comment]
@albanD (also `bias=False` naming is probably better (to match nn.Linear's arguments' naming)

**Comment by albanD** [comment]
yes `bias` is not great as it is usually just the bias weight but here it also means the centering is also removed (you don't remove the average bias). Do you think `rms_only=True` would be a better name?


> tbh, if I could just use torch.compile for one-off modules, that'd be sick. It feels like overkill to try to compile an entire model.

@PetrochukM this is the whole point of torch.compile (and the big difference with existing jit in ML frameworks) it is designed to work with partial graphs and small pieces!
What I heard from @cpuhrsch is that in the case of RMSNorm, torch.compile is able to generate a good compiled version. So wrapping just that one layer implementing it will give you a fused version.

**Comment by stas00** [comment]
Thank you for clarifying that one doesn't have to compile the whole model to fuse just one component, albanD. That's great!

I disagree that this should be left to user. You want pytorch to be the winning framework, correct? Make it great out of the box. Expecting users to figure out that they need to `compile` `RMSNorm` is not a great strategy, IMHO.

**Comment by vadimkantorov** [comment]
I tend to agree with @stas00 as pytorch core/domain libraries is an important source of idioms that are adopted my the community. So if RmsNorm module can be implemented trivially in core by torch.compiling around simple impl, it's great to have it in core + tests + perf tests. Then this could also be a showcase/doc reference of usecase where torch.compile works great (kind of dogfooding).

**Comment by PetrochukM** [comment]
Yeah, if compiling works, why not offer out-of-the-box PyTorch modules that have been compiled together? Especially, modules that are used in state-of-the-art models. 

Most state-of-the-art models cannot be built in PyTorch because they require many layers that are not readily available, and are hard to implement. The community is needing to use libraries like `apex` which are hard to use.

**Comment by ezyang** [comment]
@albanD the problem is that we need to do a zero-to-one of having torch.compile inside regular torch library code. This is a reasonable thing to want to do, but there hasn't really been any emphasis on it (since most of the effort has been on torch.compile with full libraries). Because there is no emphasis on this style of use case, there are big perf gaps (e.g., guard evaluation overhead matters a lot more in this regime).

So to make your suggestion into reality, we need to also spend some time making this work well. Or we just enlist the OSS community's help in just adding the FusedRMSNorm into the framework directly and kick the can a few more months.

**Comment by albanD** [comment]
> I disagree that this should be left to users. You want pytorch to be the winning framework, correct? Make it great out of the box.

This is definitely the long term vision where compile will be always there, optimizing what it can and leaving the rest as python code.
In that world, users won't have to figure out they need compile. They just always use it and will get the perf out of the box.

Right now we're in a weird transition period though as, as you said, this is not stable enough to be in this always-on state. And so the question is still open on how to best spend our ressources: add more fused one-off kernels that will be obsolete when compile is the default or work on making compile the default.
I think our current stance is to still add these fused kernels for things that are critical right now or that compile will not handle well any time soon (SDPA, adamw for example) and focus the rest of our efforts to making compile the default.

To come back to this particular issue, adding the new flag to be able to do RMSNorm is a definite "yes we want it" but adding the fused implementation is more "we would accept a simple PR but no one on the core team is working on it".

**Comment by stas00** [comment]
If things are unstable, yet, it'd be wasteful to allocate resources to do such porting because an automatic solution is "imminent", then I think it's perfectly fine to close this feature request and tell the user to use `apex` if they want speed - a user who can figure out how to compile `RMSNorm` will surely be able to build `apex`, so I don't really see any advantage here if it's not something available out of the box.

**Comment by ZixuanJiang** [comment]
We **unify** the `LayerNorm` and `RMSNorm` in Pre-Normalization Transformers in our paper https://arxiv.org/abs/2305.14858. The arithmetic equivalence allows us to convert Pre-LayerNorm Transformers into Pre-RMSNorm models without impact on the model functionality.

Considering that RMSNorm offers superior efficiency compared to LayerNorm in theory, we believe that providing an official RMSNorm API would greatly benefit the community, allowing them to harness this improvement in both training and inference effectively.

We also release our implementation https://github.com/ZixuanJiang/pre-rmsnorm-transformer for reference. Thanks for your consideration.

**Comment by mikaylagawarecki** [comment]
Following up on the discussion here, it seems like

`LayerNorm` is given by 

$y = \frac{x - \mathrm{E}[x]}{ \sqrt{\mathrm{Var}[x] + \epsilon}} * \gamma + \beta$
where the $\epsilon$ is added for numerical stability

and `RMSNorm` is given by 

$y = \frac{x}{ RMS(x)} * \gamma$ (with maybe an optional $+  \beta$ [per this comment](https://github.com/pytorch/pytorch/issues/72643#issuecomment-1035594258)) 

Where 

$Var(x) = \frac{1}{n}\sum\limits_{i=1}^{n}  (x_i - E[x])^2$

and 

$RMS(x) =  \sqrt{\frac{1}{n}\sum\limits_{i=1}^{n}   (x_i)^2}$ ($\epsilon$ could also be used for numerical stability in this calculation)

so `RMSNorm` defers from `LayerNorm` on 3 counts 
1) don't subtract expectation from numerator
2) use `RMS(x)` rather than $\sqrt{Var(x) + \epsilon}$ in the denominator
3) (perhaps) don't learn an elementwise affine bias 

If my understanding here is correct, we would accept the addition of `RMSNorm` as a new `torch.nn.Module` that is separate from `LayerNorm`

**Comment by vadimkantorov** [comment]
It also seems that there's a RMSNorm impl in flash-attention (fused with dropout): https://github.com/Dao-AILab/flash-attention/blob/4f285b354796fb17df8636485b9a04df3ebbb7dc/flash_attn/ops/rms_norm.py#L11

**Comment by vadimkantorov** [comment]
Am I also understanding correctly that semantically RMSNorm(x) = F.normalize(x, dim = dim, p = 2) * sqrt(D) * gamma + beta?

**Comment by vadimkantorov** [comment]
There is also an impl of RMSNorm in Triton at https://github.com/kakaobrain/trident/blob/main/trident/kernel/rms_norm.py - maybe can be incorporated into core?

**Comment by Ryu1845** [comment]
Any progress on this issue ? cc: @albanD

**Comment by vince62s** [comment]
as a matter of fact the rmsnorm from FasterTransformer is much faster than Apex'
https://github.com/NVIDIA/FasterTransformer/blob/main/src/fastertransformer/kernels/layernorm_kernels.cu

**Comment by stas00** [comment]
Could you please share some factual information to support this claim, @vince62s?

**Comment by vince62s** [comment]
@stas00 I can try to make a unitary snippet but can tell you in this PR https://github.com/OpenNMT/OpenNMT-py/pull/2539/files it had a huge impact on inference tok/sec for a Mistral7B LM for instance. (separately from the other change kv_cache from flash2 that had also an impact).

**Comment by vadimkantorov** [comment]
Related: https://github.com/pytorch/pytorch/issues/121526

**Comment by Ryu1845** [comment]
Should probably be closed now

**Comment by albanD** [comment]
The PR adds the API but not the fused kernel. So we can keep this open in case someone wants to investigate if we would get benefit from a manually fused kernel. And if so, wants to upstream such implementation.

**Comment by vadimkantorov** [comment]
Related, I guess: https://github.com/pytorch-labs/applied-ai/blob/main/kernels/triton/training/rms_norm/fused_rms_norm.py by @lessw2020

**Comment by GuWei007** [comment]
https://github.com/huggingface/transformers/issues/30236
When I was fine-tuning the llama model, a loss spike occurred. I changed the mul in RMSnorm to fp32, which solved the problem. Can you help me look at the strategy for improving the computing dtype  of the RMSnorm operator?

**Comment by vadimkantorov** [comment]
This is quite interesting, maybe there needs to be some more expressive torch.mul args to allow some upcasts during the multiplication itself, e.g. `torch.mul(a, b, compute_dtype = torch.float32, dtype = torch.bfloat16)` or `torch.mul(a, b, upcast = True, out_dtype = torch.bfloat16)`

**Comment by vadimkantorov** [comment]
It also appears that Intel IPEX has some optimized code paths for RMSNorm and LayerNorm for CPUs: https://github.com/intel/intel-extension-for-pytorch/tree/main/examples/cpu/inference/python/llm-modeling

**Comment by vadimkantorov** [comment]
Another fused RMSNorm impl in Triton: https://x.com/hsu_byron/status/1827072742291861975 at https://github.com/linkedin/Liger-Kernel

@msaroufim The max mem improvements in Liger are massive. Curious if some of these optimizations could be added to core (like Linear layer + softmax + crossentropy. Inductor would not create such fusions now?) or if these Triton kernels could be added to PyTorch (at least as some known fusion patterns for Inductor). And of course the best would be to be able to generate these fusions with Inductor

**Comment by msaroufim** [comment]
So for this kind of work we basically have 4 options sorted in decreasing difficulty
1. Have inductor code generate the kernel: ideal but not possible when the kernel has clever algorithmic rewrites
2. Pattern match: a compromise for algorithmic rewrites 
3. Package the kernel in core: easy because it's pure python but hard because we've tended to be conservative on which kinds of kernels we can add
4. Package the kernel out of core: this is the easiest to do, whether its in AO or elsewhere 

Honestly I'd just vote for 4 since that's easiest and useful now. 3 isn't that crazy nowadays since it's easier to package Triton kernels since they get JIT'd vs CUDA. For 1 and 2 it sounds like the right long term thing to do but would love to hear more from @Chillee or @eellison on this, how hard is it go from a solid Triton kernel to getting it working with the pattern matcher today, ideally if there's a few n00b friendly reference examples it could help us move faster here

**Comment by vadimkantorov** [comment]
I think the main advantages of in-core is discoverability and resources to keep it tested for new accelerators and be notified if any regressions happen (on a large pytorch's bench) and avoiding forcing other packages to take on another dependency. Maybe if torchao can be a "staging" area before inclusion in core, but probably main packages like HF could be wary of taking a dependency on it too (or maybe not?)

The problem with 4 is that it keeps the speeding-up packages fragmented and often not tested/plugging into each other. E.g. another speeding-up packages announced yesterday: https://x.com/DAlistarh/status/1826538436225806468. Currently the wide adoption is happening via domination of either PyTorch core or some other ultra-popular packages like HF.

At the end, IMO everyone will win if good continuously tested triton kernels somehow could be widely adopted - either by PyTorch core (eventually) or via some other ultra-popular package (I hope it was torchao, but I think for now it's just one of tons other packages experimenting with quantization APIs and kernels)

**Comment by Chillee** [comment]
From my testing, I remember that vanilla torch.compile has on-par/better perf compared to the handwritten triton kernels.

**Comment by vadimkantorov** [comment]
Maybe the only interesting thing there is the fused linear + softmax + loss (in terms of max mem reduction). Or another question is why it's showing such big speedups against HF, assuming HF can now use torch.compile... Or is the problem that torch.compile does not decompose / compile RMSNorm / applying RoPE into Triton?

**Comment by Chillee** [comment]
The chunked linear + crossentropy makes a big difference in terms of peak memory, and for finetuning where you're often batch-size constrained, I've seen it make a huge difference in terms of total throughput.

**Comment by vadimkantorov** [comment]
So I guess, maybe what could be cool is inclusion in PyTorch core of some exteremely popular shortcuts from popular LLM models (like RoPE / SwiGLU) and having torch.compile applied to them by PyTorch itself. At least it would raise the eager baseline for any new proposed fused ops...

**Comment by vadimkantorov** [comment]
Unsloth also includes Triton kernels for LayerNorm and RMSNorm (apache license):
- https://github.com/unslothai/unsloth/blob/main/unsloth/kernels/layernorm.py
- https://github.com/unslothai/unsloth/blob/main/unsloth/kernels/rms_layernorm.py (apache license)

I wonder if this means that PyTorch native torch.compile inductor/triton codegen does not produce fast enough code for LayerNorm/RMSNorm?
